### ISSUE TITLE
Parsing and outputting corrupts the apk

### ISSUE THREAD

**Issue by bichselb** [soot-oss/soot#644]
For some apks, if I parse them and write them back again, this results in a corrupt apk. I realize this because dex2jar runs into an OutOfMemoryError for the new apk.

Steps to reproduce this problem:
1. Download e.g. https://play.google.com/store/apps/details?id=com.antivirus using http://apk-dl.com
2. Run the code shown below (using Soot from latest nightly build at https://ssebuild.cased.de/nightly/soot/lib/soot-trunk.jar )
3. Run dex2jar 2.0 on the produced apk
   `d2j-dex2jar.sh AntiVirus_FREE_2016_-_Android_5.7_apk-dl.com.apk`
4. Obtain an exception

**Exception**

```
Exception in thread "main" java.lang.OutOfMemoryError: GC overhead limit exceeded
    at com.googlecode.dex2jar.ir.stmt.Stmts.nGoto(Stmts.java:32)
    at com.googlecode.dex2jar.ir.ts.EndRemover.move4End(EndRemover.java:96)
    at com.googlecode.dex2jar.ir.ts.EndRemover.transform(EndRemover.java:59)
    at com.googlecode.d2j.dex.Dex2jar$2.optimize(Dex2jar.java:134)
    at com.googlecode.d2j.dex.Dex2Asm.convertCode(Dex2Asm.java:414)
    at com.googlecode.d2j.dex.ExDex2Asm.convertCode(ExDex2Asm.java:42)
    at com.googlecode.d2j.dex.Dex2jar$2.convertCode(Dex2jar.java:128)
    at com.googlecode.d2j.dex.Dex2Asm.convertMethod(Dex2Asm.java:509)
    at com.googlecode.d2j.dex.Dex2Asm.convertClass(Dex2Asm.java:406)
    at com.googlecode.d2j.dex.Dex2Asm.convertDex(Dex2Asm.java:422)
    at com.googlecode.d2j.dex.Dex2jar.doTranslate(Dex2jar.java:172)
    at com.googlecode.d2j.dex.Dex2jar.to(Dex2jar.java:272)
    at com.googlecode.dex2jar.tools.Dex2jarCmd.doCommandLine(Dex2jarCmd.java:108)
    at com.googlecode.dex2jar.tools.BaseCmd.doMain(BaseCmd.java:288)
    at com.googlecode.dex2jar.tools.Dex2jarCmd.main(Dex2jarCmd.java:32)
```

**Code**

```
public class Main {

    public static void main(String[] args) {
        Main m = new Main();
        m.initialize();
        m.printCurrentSootContent();
    }

    public void initialize() {
        Options.v().set_no_bodies_for_excluded(true);
        Options.v().set_allow_phantom_refs(true);
        Options.v().set_whole_program(true);
        Options.v().set_process_dir(
                Collections.singletonList("./apk/in/AntiVirus_FREE_2016_-_Android_5.7_apk-dl.com.apk"));
        Options.v().set_android_jars("./libs/android-platforms");
        Options.v().set_src_prec(Options.src_prec_apk);
        soot.Main.v().autoSetOptions();
        Scene.v().loadNecessaryClasses();
    }

    public void printCurrentSootContent() {
        Options.v().set_output_format(soot.options.Options.output_format_dex);
        Options.v().set_output_dir("./apk/out");
        PackManager.v().writeOutput();
    }
}
```

**Comment by StevenArzt** [comment]
Hi,

Can you please double-check that this is really an issue with Soot and 
not with dex2jar? Have you tried to run the APK files written by Soot on 
an emulator? Have you tried to run them through Soot once again? If that 
all works, then the issue lies with dex2jar. Otherwise, we need to have 
a closer look at what's going on inside Soot.

Best regards,
   Steven

**Comment by bichselb** [comment]
Hi

Thanks for your incredibly fast response!

If I remember correctly, Soot does not have a problem parsing the produced apk again.

However, I have ran the produced apk on an emulator to provoke an exception. This exception does not occur on the original apk.


The steps I have performed:

- Sign the generated apk
```
keytool -genkey -v -keystore my-release-key.keystore -alias alias_name -keyalg RSA -keysize 2048 -validity 10000
jarsigner -verbose -sigalg SHA1withRSA -digestalg SHA1 -keystore my-release-key.keystore ./AntiVirus_FREE_2016_-_Android_5.7_apk-dl.com.apk alias_name
```
- Run Android Emulator from Android Studio 2.2 (Nexus 4 API 24, Android 7.0)
- Install the produced apk
```
adb install -r ./AntiVirus_FREE_2016_-_Android_5.7_apk-dl.com.apk
```
- Open the app on the emulator (the installation works)
- Click around: "GET STARTED" -> "CONTINUE WITH ADDS" -> "SCAN"
- Receive the error "AntiVirus has stopped"
